apps/test/unit/lib/kits/maker/boards/microBit/MicroBitMakerBoardTest.js

```diff
@@ -0,0 +1,153 @@
+    'MicroBitThermometer'
+  ];
+
+  /**
+   * After installing on the interpreter, test that the components and
+   * component constructors are available from the interpreter
+   */
+  describe('Micro Bit components accessible from interpreter', () => {
+    let jsInterpreter;
+    let board;
+
+    beforeEach(() => {
+      board = new BoardClient();
+      if (boardSpecificSetup) {
+        boardSpecificSetup(board);
+      }
+
+      jsInterpreter = {
+        globalProperties: {},
+        createGlobalProperty: function(key, value) {
+          jsInterpreter.globalProperties[key] = value;
+        },
+        addCustomMarshalObject: sinon.spy()
+      };
+
+      return board.connect();
+    });
+
+    describe('adds component constructors', () => {
+      beforeEach(() => {
+        board.installOnInterpreter(jsInterpreter);
+      });
+
+      it(`correct number of them`, () => {
+        expect(jsInterpreter.addCustomMarshalObject).to.have.callCount(
+          MB_CONSTRUCTOR_COUNT
+        );
+      });
+
+      MB_COMPONENTS.forEach(constructor => {
+        it(constructor, () => {
+          expect(jsInterpreter.globalProperties).to.have.ownProperty(
+            constructor
+          );
+          expect(jsInterpreter.globalProperties[constructor]).to.be.a(
+            'function'
+          );
+          const passedObjects = jsInterpreter.addCustomMarshalObject.args.map(
+            call => call[0].instance
+          );
+          expect(passedObjects).to.include(
+            jsInterpreter.globalProperties[constructor]
+          );
+        });
+      });
+    });
+
+    describe('adds components', () => {
+      beforeEach(() => {
+        board.installOnInterpreter(jsInterpreter);
+      });
+
+      it(`correct number of them`, () => {
+        let globalPropsCount = MB_CONSTRUCTOR_COUNT + MB_COMPONENT_COUNT;
+        expect(Object.keys(jsInterpreter.globalProperties)).to.have.length(
+          globalPropsCount
+        );
+      });
+
+      ['buttonA', 'buttonB'].forEach(button => {
+        describe(button, () => {
+          let component;
+
+          beforeEach(() => {
+            component = jsInterpreter.globalProperties[button];
+          });
+
+          it('isPressed', () => expect(component.isPressed).to.be.a('boolean'));
+          it('holdtime', () => expect(component.holdtime).to.be.a('number'));
+        });
+      });
+
+      describe('ledMatrix', () => {
+        function expectLedToHaveFunction(fnName) {
+          expect(jsInterpreter.globalProperties.ledMatrix[fnName]).to.be.a(
+            'function'
+          );
+        }
+
+        // Set of required functions derived from our dropletConfig
+        [
+          'on',
+          'off',
+          'toggle',
+          'clear',
+          'scrollString',
+          'scrollNumber'
+        ].forEach(fnName => {
+          it(`${fnName}()`, () => expectLedToHaveFunction(fnName));
+        });
+      });
+
+      describe('tempSensor', () => {
```

Comment:
I noticed these last few tests are duplicates. Is that intentional?

Reply:
Yup! This is a decision I would like feedback on. I decided to duplicate these tests to draw a firm line of: "If it's part of the board, it should be in that board's test". The other option is to have three separate tests - MB, CP, shared.

Reply:
I really like the approach you've taken here where the tests for each board include a complete component list.

Here's an unreasonably long answer to my thought process on whether to refactor this. :grin:

If you want to completely DRY up the code, you could compose your tests by extracting the describe blocks for some board components and building a small vocabulary of rich assertions.

```js
describe('ledMatrix', () => {
  // ...
});

assertTempSensorOnInterpreter();

assertAccelerometerOnInterpreter();

assertBoardOnInterpreter();

// In some shared file, like makerComponentAssertions.js
function assertTempSensorOnInterpreter() {
  describe('tempSensor', () => {
    // ...your existing test code here
  });
}
```

That said, I'd first consider a few reasons _not_ to do this yet:

1. **Is this false duplication?** Code that is identical _now_ is not duplicated if it changes at different rates over time.  If you DRY up false duplication, you're just creating future work for yourself splitting it apart again.
   In this case I think we have both a true duplication and a false one.
   The false duplication is in the list of components for each board: Given component lists `CP: [A, B, C, D, E]` and `MB: [A, B, C, F]` I don't think we need to somehow deduplicate the fact that ABC are shared in both lists.
   The true duplication is in the tests of interface details for these component types: The fact that `tempSensor` has `F` and `C` properties is likely to remain true for all board types, and/or change for all board types at once.
   
2. **Have you satisfied the [rule of three](https://stackoverflow.com/a/2298357/5000129)?** Not everybody agrees with this heuristic but it's popular and has been studied, somewhat.  I tend to think of this in terms of maintenance cost: Abstractions are powerful, but they usually add complexity - they're a little harder to read and a little harder to reason about.  The cost/benefit when using an abstraction to deduplicate code often tips at 3 cases.  You can always deduplicate later, _if_ you notice that you've already got two cases.

3. **[YAGNI](https://martinfowler.com/bliki/Yagni.html)?** I might not wait for the rule of three if it was obvious that a third case was going to exist soon.  If not, maybe we should save ourselves the effort.  In this case, even though we'd _like_ to see support for a third board type in our near future, I'm not aware of any such plans.  It seems to me the risk of extracting these helper assertions and then never using them again is fairly high right now.

Given those three heuristics, I think you're justified in leaving this duplication in place for now.  I'd only move forward with deduplication if a) you thought it could significantly improve the _readability_ of these tests, or b) you're interested in trying it, just as an exercise.

Reply:
Based on this feedback, I think I'm going to stick with this duplication as is.